jvm-packages/python/sparkxgb/xgboost.py

```diff
@@ -0,0 +1,254 @@
+            if not hasattr(self, java_param_name):
+                param = Param(self, java_param_name, java_param.doc())
+                setattr(param, "created_from_java_param", True)
+                setattr(self, java_param_name, param)
+                self._params = None  # need to reset so self.params will discover new params
+
+    # Backport SPARK-10931 (For Spark <2.3)
+    def _transfer_params_from_java(self):
+        """
+        Transforms the embedded params from the companion Java object.
+        """
+        sc = SparkContext._active_spark_context
+        for param in self.params:
+            if self._java_obj.hasParam(param.name):
+                java_param = self._java_obj.getParam(param.name)
+                # SPARK-14931: Only check set params back to avoid default params mismatch.
+                if self._java_obj.isSet(java_param):
+                    value = _java2py(sc, self._java_obj.getOrDefault(java_param))
+                    self._set(**{param.name: value})
+                # SPARK-10931: Temporary fix for params that have a default in Java
+                if self._java_obj.hasDefault(java_param) and not self.isDefined(param):
+                    value = _java2py(sc, self._java_obj.getDefault(java_param)).get()
+                    self._setDefault(**{param.name: value})
+
+    # Override the "_from_java" method, so we can read our objects.
+    @classmethod
+    def _from_java(cls, java_stage):
+        """
+        Given a Java object, create and return a Python wrapper of it.
+        """
+
+        # Create a new instance of this stage.
+        py_stage = cls()
+
+        # Load information from java_stage to the instance.
+        py_stage._java_obj = java_stage
+        py_stage._create_params_from_java()
+        py_stage._resetUid(java_stage.uid())
+        py_stage._transfer_params_from_java()
+
+        return py_stage
+
+
+class XGBoostEstimator(JavaParamsOverrides, JavaEstimator, HasCheckpointInterval, HasFeaturesCol, HasLabelCol,
+                       HasPredictionCol, HasWeightCol, JavaMLWritable, XGBoostReadable):
+    """
+    A PySpark implementation of ml.dmlc.xgboost4j.scala.spark.XGBoostEstimator.
+    """
+
+    @keyword_only
+    def __init__(self,
+                 # General Params
+                 checkpoint_path="", checkpointInterval=-1, missing=None, nthread=1, nworkers=1, silent=0,
+                 use_external_memory=False,
+
+                 # Column Params
+                 baseMarginCol="baseMargin", featuresCol="features", labelCol="label", predictionCol="prediction",
+                 weightCol="weight",
+
+                 # Booster Params
+                 base_score=0.5, booster="gbtree", eval_metric="error", num_class=2, num_round=2,
+                 objective="binary:logistic", seed=None,
+
+                 # Tree Booster Params
+                 alpha=0.0, colsample_bytree=1.0, colsample_bylevel=1.0, eta=0.3, gamma=0.0, grow_policy='depthwise',
```

Comment:
Although the  grouped  category parameters seems easier for us, maybe  it's better to make  them a litter compact because they are class init arguments. Please correct me if I'm wrong.